### ISSUE TITLE
Using io.sails with JWT Bearer Tokens

### ISSUE THREAD

**Issue by sgentile** [balderdashy/sails#2759]
inside angular client : 
var socket = io.sails.connect();
        socket.get('/user/' + currentUser().id + '/messages', function(model) {
            // do whatever you want with it.
        });

I get an expected result with using err: "No Authorization header was found" when using sockets with JWT bearer tokens:

Surely this has to be a common scenario right ?  ie. https://auth0.com/blog/2014/01/15/auth-with-socket-io/

I'm not sure how to go about doing this with Sails (v0.11.0) ?

(note, I'm using this as my example starting place:  https://github.com/Foxandxss/sails-angular-jwt-example )

**Comment by weblogixx** [comment]
@sgentile: I had a similiar issue. I use sails 0.11.0 together with Angular using the io.socket.connect functionality. The problem is, that the bearer token needs to be send along the request. You need to set a config variable (io.sails.query = 'token=MYSUPERTOKEN') for this to work.

Also, due to a bug in  sails.io.js it does not work without a modification. See https://github.com/balderdashy/sails.io.js/pull/46.

I have already pushed a patch for this to work, but it is not in there yet. In the meantime, just open up your socket.io.js file and modify it in the following way:

```
    function SailsSocket (opts){
      var self = this;
      opts = opts||{};

      // Absorb opts
      self.useCORSRouteToGetCookie = opts.useCORSRouteToGetCookie;
      self.url = opts.url;
      self.multiplex = opts.multiplex;
      self.transports = opts.transports;

      // Add this
      self.query = opts.query;
```

After this change, it should work correctly. Hope this helps!

**Comment by sgentile** [comment]
Thank you..  I tried some things but still not working:

var currentUser = CurrentUser.user;

        var token = angular.fromJson(LocalService.get('auth_token')).token;
        io.sails.query = 'token=' + token;
        var socket = io.sails.connect();
        socket.get('/user/' + currentUser().id + '/messages', function(a, b) {
            // do whatever you want with it.
            console.log('get', a);
            debugger;
        });

I added the sails socket code as you instructed.

Problem I'm seeing here is connect is not picking up those opts ie.  I show a 'null'
io.sails.connect = function(url, opts) {
      opts = opts || {};

      // If explicit connection url is specified, save it to options
      opts.url = url || opts.url || undefined;

      // Instantiate and return a new SailsSocket- and try to connect immediately.
      var socket = new SailsSocket(opts);
      socket._connect();
      return socket;
    };

But when I look at the io.sails object - it has the query object in it:
query: "token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzaWQiOjEsImlhdCI6MTQyNjg1MTI4Nn0.TXnvffKUQc3PQADmLeG1VnIvz8b2IBetDvemp6XpT6A"

I tried changing from self.query  = opts.query (as opts was always null) to self.query = io.sails.query;

And I see that token coming through - but I still get the error :
{err: "No Authorization header was found"}

**Comment by weblogixx** [comment]
This looks quite good. What I did was not using the io.sails.connect method, but adding a custom policy that handles REST as well as the Socket.IO API. If you want I could send an example tomorrow.

**Comment by sgentile** [comment]
That would be fantastic - thanks!

**Comment by weblogixx** [comment]
@sgentile My code is a heavily modified one, based on this tutorial: http://angular-tips.com/blog/2014/05/json-web-tokens-examples/. Just go through it step by step for the sails side and it should work.

Just update the policiy:

```
// Replace:
if (req.headers && req.headers.authorization) {

// with:
  // If we have a socket connection, use its query token if available
  if(req.isSocket) {
      if(req.socket.handshake && req.socket.handshake.query && req.socket.handshake.query.token) {
          token = req.socket.handshake.query.token;
      }
  } else if (req.headers && req.headers.authorization) {
```

Hope this helps.

**Comment by sgentile** [comment]
That helps.  So, I've boiled down the issue to the sails.io.js piece as above - with this code here you shared it works - but I have to manually set the self.query - the opts.query isn't passed along
// Add this
      self.query = opts.query;

This alone doesn't appear to work:

var token = angular.fromJson(LocalService.get('auth_token')).token;
        io.sails.query = 'token=' + token;
        var socket = io.sails.connect();
        socket.get('/user/1/messages', function(a, b) {
            // do whatever you want with it.
            console.log('get', a, b);
        });

Instead sails.connect the io.sails.query variable isn't being set

I guess my next piece to understand is why this isn't happening.   Thanks again for your help, I think I'm getting closer.

**Comment by weblogixx** [comment]
Could you please try to use https://raw.githubusercontent.com/weblogixx/sails.io.js/master/sails.io.js? Seems like I changed one more line (and a plugin changed some formatting too :(). The other line (in the connect method) is currently missing in your file.

Hope this helps.

**Comment by billyshena** [comment]
Is there a way to add an interceptor (middleware) on the io.sails requests? I would like to catch for example all the 401, 403 returned by my API server.


Best,

**Comment by pcguru** [comment]
I went for just monkey-patching get, post, put, delete actions to always send the header over sockets to sailsjs, since that is the only header I need right now.

	io.socket.get = function(url, data, cb) {
	  // `data` is optional
	  if (typeof data === 'function') {
		cb = data;
		data = {};
	  }

	  return this.request({
		method: 'get',
		params: data,
		url: url,
		headers: {
		  authorization: 'Bearer ' + getToken()
		}
	  }, cb);
	};

It would be nice if we could set up some default headers to be sent, like jQuery does it for example.
Something like:

    io.setDefaults({
        headers: {
            Authorization: 'Bearer ' + getToken()
        }
    });

**Comment by sailsbot** [comment]
Thanks for posting, @sgentile.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by michaek** [comment]
@pcguru Do you think it might be cleaner to wrap io.socket.request instead of patching each http method function?

```
var originalRequest = io.socket.request.bind(io.socket);
io.socket.request = function (options, cb) {
	options.headers = options.headers || {};
	options.headers['Authorization'] = 'Bearer ' + getToken();
	originalRequest(options, cb);
};
```
